lib/ansible/modules/system/osx_defaults.py

```diff
@@ -15,413 +16,975 @@
+    - List and dict can't be given as value in key=argument form.
+    - Binary data need to be encoded in base64.
 '''
 
 EXAMPLES = '''
+# Show debug menu in Safari. Specifying the type.
 - osx_defaults:
     domain: com.apple.Safari
     key: IncludeInternalDebugMenu
     type: bool
     value: true
     state: present
-
+# Set Measurement unit to cm for all apps (NSGlobalDomain).
 - osx_defaults:
     domain: NSGlobalDomain
     key: AppleMeasurementUnits
     type: string
     value: Centimeters
     state: present
-
+# Set clock visibility on screensaver for all users on current host.
 - osx_defaults:
     domain: com.apple.screensaver
     host: currentHost
+    any_user: true
     key: showClock
     type: int
     value: 1
-
+# Set Measurement unit to cm for all apps (NSGlobalDomain). No type specified.
 - osx_defaults:
     key: AppleMeasurementUnits
     type: string
     value: Centimeters
-
+# Set AppleLanguages to a list of languages.
 - osx_defaults:
     key: AppleLanguages
-    type: array
-    value:
-      - en
-      - nl
-
+    value: ["en", "nl"]
+# Remove a key.
 - osx_defaults:
     domain: com.geekchimp.macable
     key: ExampleKeyToRemove
     state: absent
+# Set a date. Dates need to be quoted. Timezone can be omitted. Local timezone will be use in this case.
+- osx_defaults:
+    domain: com.geekchimp.macable
+    key: SomeDate
+    value: "2002-12-15 02:59:43Z"
+# Set the nested key arrangeBy in abbreviated form (nested key need to be quoted).
+- osx_defaults: { domain: com.apple.finder, key: 'DesktopViewSettings:IconViewSettings:arrangeBy', value: dateModified }
+# Set ListViewSettings key to complex nested values.
+- osx_defaults:
+    domain: com.apple.finder
+    key: ComputerViewSettings:ListViewSettings
+    value:
+        - iconSize: 16
+        - sortColumn: name
+        - textSize: 12
+        - columns:
+            - comments:
+                - ascending: true
 '''
 
+
+from base64 import b64decode
+import calendar
+import contextlib
 import datetime
+import os
 import re
+import string
 
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.six import binary_type, text_type
-
-
-# exceptions --------------------------------------------------------------- {{{
-class OSXDefaultsException(Exception):
-    pass
-
-
-# /exceptions -------------------------------------------------------------- }}}
-
-# class MacDefaults -------------------------------------------------------- {{{
-class OSXDefaults(object):
-    """ Class to manage Mac OS user defaults """
+from ansible.module_utils.six import string_types, binary_type, text_type
+from ansible.module_utils.six import integer_types
+
+try:
+    import CoreFoundation
+    import Foundation
+    has_foundation = True
+except ImportError:
+    has_foundation = False
+
+
+# Regular expression that match datetime formats. Should match mostly the
+# same formats that YAML support.
+RE_DATETIME = re.compile(r"""
```

Comment:
The main reason why I've used a regex is because it's complicated to [support the timezone with `strptime` in python 2.7](https://stackoverflow.com/a/38992623/146481) (this is just an example, we also need to supports 'Z', etc.).

It gives also a bit of flexibility on the format. I should mention that this regex is adapted from the one defined in the [YAML specifications](http://yaml.org/type/timestamp.html), so it should be pretty safe and it's not really complex. It is also virtually conforming to the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format, a format that gives this flexibility.

About the partial date, if you mean a date without a timezone (a naive date), my code add the local timezone when there's no timezone. This mimics how the `defaults` command works.

Reply:
Fair enough, I just tend to be careful about complex regex'es. Maybe then link to their page too, might be useful for someone else who reads to know why the regex was picked?

Reply:
Yes, I should add the references.